### DISCUSSION TITLE
Adding audio track to PeerConnection

### DISCUSSION THREAD

**Discussion by yrucrem** [paullouisageneau/libdatachannel#1244] [Q&A] [answered]
Hello!

I'm trying to send audio data over a peer connection and I'm not sure that my setup for that is correct. Both sides of the connection use libdatachannel for WebRTC. Establishing the peer connection works and I can send data over the datachannel, but when I add an audio track and then try to send data via Track::send() I just get warnings on the server side:

```
Number of RTP packets received with an unknown packet type: 61 (over 1 seconds)
```

When initializing the peer connection I made sure to set `forceMediaTransport` to `true` and I added an onTrack callback:

```
	peer_connection->onTrack([this](std::shared_ptr<rtc::Track> track) {
		printf("Track added\n");
		// ...
	});
```

The code to add the track looks like this:

```
	rtc::Description::Audio audio("audio", rtc::Description::Direction::SendOnly);

	audio.addSSRC(next_ssrc++, "audio-send");

	std::shared_ptr<rtc::Track> track = peer_connection->addTrack(audio);
```

where `next_ssrc` is a running counter. The created track is put into a map and when audio data is available I create a std::byte array and send it with

```
	track->send(data, data_len);
```

I noticed that I do not see the "Track added" message on the remote peer connection, so I'm not sure if the remote peer is even aware that the track was added. And when I start sending data to the track, I get the messages about unknown packet types on the other end.

Am I missing a step in setting up the connection?

Any hints and tipps on how to establish an audio connection between two libdatachannel peer connections would be greatly appreciated!

Thanks in advance and best regards,
Patrick

**Comment by paullouisageneau** [comment]
There are a few things missing: 
- You need to declare the format of your audio data with an RTP payload type in the description. For instance, if it is Opus-encoded:
```
audio.addOpusCodec(96);
```
- If your audio stream is not already packetized as RTP, you must configure the track to packetize your audio stream. For instance, for Opus:
```
auto rtpConfig = make_shared<RtpPacketizationConfig>(ssrc, "audio-send", 96);
auto packetizer = make_shared<OpusRtpPacketizer>(rtpConfig);
track->setMediaHandler(packetizer);
```
- After each send, you must increase `rtpConfig->timestamp` by the audio duration in samples. You should send audio in chunks matching the audio frame duration (20ms by default), so for Opus at 48000Hz in would be an increase of 960. Alternatively you can also use `RtpPacketizationConfig::secondsToTimestamp()` helper function. (I'm planning to introduce a `sendFrame()` method that would do that for you provided you send the expected duration.)

**Reply by yrucrem** [reply]
Hello!

Thank you for your input @paullouisageneau! That sounds very promising, I will try those changes as soon as I can.

Best regards,
Patrick

**Reply by yrucrem** [reply]
Hello again!

First of all, again a big thanks, that really helped me a lot! The warnings about the unknown packet type on the server no long appear.

What still does not work unfortunatelly is that the onTrack() handler is called server side. I call onTrack() on the newly created peer connection and have a logging message at the beginning, but I never see the message in the output.

```
	peer_connection->onTrack([this](std::shared_ptr<rtc::Track> track) {
		printf("Track added\n");
		// ...
	});
```

Am I understanding it correctly that I need that reference to the track on the server side, so that I can add an onMessage() handler to do something with the data?

If this is not the intended way, could you please elaborate how the setup works to receive the audio data on the other end of the connection?

Thanks in advance and best regards,
Patrick

**Reply by paullouisageneau** [reply]
If `OnTrack` is not triggered, it means the no offer containing the track was received.

If you add the track after creating a data channel, be aware that by default `createDataChannel()` initiate the negotiation (equivalent to calling `setLocalDescription()`), so you must either create tracks before data channels or set `disableAutoNeogotiation` in the config and call `setLocalDescription()` yourself.

**Reply by yrucrem** [reply]
Hello and thanks for your continued help on this issue!

I double checked that negotiation is not started before adding the audio track, but the onTrack() callback still is not called.

The `disableAutoNeogotiation` setting is `true` and `setLocalDescription()` is called after the audio track is added and the datachannel is created.

This is the snipped creating the peer connection:

```
	r_config.disableAutoNegotiation = true;

	peer_connection = std::make_shared<rtc::PeerConnection>(r_config);

	peer_connection->onLocalDescription([this](rtc::Description description) {
		std::cout << "SESSION DESCRIPTION CREATED" << std::endl;
		// ...
	});
	peer_connection->onLocalCandidate([this](rtc::Candidate candidate) {
		std::cout << "ICE CANDIDATE CREATED" << std::endl;
		// ...
	});
	peer_connection->onDataChannel([this](std::shared_ptr<rtc::DataChannel> channel) {
		std::cout << "DATACHANNEL CREATED" << std::endl;
		// ...
	});
	peer_connection->onTrack([this](std::shared_ptr<rtc::Track> track) {
		std::cout << "TRACK ADDED" << std::endl;
		// ...
	});
```

I noticed that only the messages for calbacks concerning the local peer connection (`onLocalDescription()` and `onLocalCandidate()`) are printed.

This is the offer that is generated on the server and if I understand correctly it _does_ contain an entry for the audio track:

```
v=0
o=rtc 2595208623 0 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE audio 0
a=group:LS audio
a=msid-semantic:WMS *
a=setup:actpass
a=ice-ufrag:DXCQ
a=ice-pwd:[ICEPWD]
a=ice-options:ice2,trickle
a=fingerprint:sha-256 [FINGERPRINT]
m=audio 9 UDP/TLS/RTP/SAVPF 96
c=IN IP4 0.0.0.0
a=mid:audio
a=sendrecv
a=ssrc:2 cname:audio-stream
a=rtcp-mux
a=rtpmap:96 opus/48000/2
a=fmtp:96 minptime=10;maxaveragebitrate=96000;stereo=1;sprop-stereo=1;useinbandfec=1
m=application 9 UDP/DTLS/SCTP webrtc-datachannel
c=IN IP4 0.0.0.0
a=mid:0
a=sendrecv
a=sctp-port:5000
a=max-message-size:262144
```

This is the answer generated by the client:

```
v=0
o=rtc 889034532 0 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE audio 0
a=group:LS audio
a=msid-semantic:WMS *
a=setup:active
a=ice-ufrag:krMJ
a=ice-pwd:[ICEPWD]
a=ice-options:ice2,trickle
a=fingerprint:sha-256 [FINGERPRINT]
m=audio 9 UDP/TLS/RTP/SAVPF 96
c=IN IP4 0.0.0.0
a=mid:audio
a=sendrecv
a=rtcp-mux
a=rtpmap:96 opus/48000/2
a=fmtp:96 minptime=10;maxaveragebitrate=96000;stereo=1;sprop-stereo=1;useinbandfec=1
m=application 9 UDP/DTLS/SCTP webrtc-datachannel
c=IN IP4 0.0.0.0
a=mid:0
a=sendrecv
a=sctp-port:5000
a=max-message-size:262144
```

And the generated ICE candidates at least contain `media=audio`:

```
ICE candidate created for 382789537: media=audio idx=0 name=candidate:2 1 UDP 2130706175 [IP6] [PORT] typ host
ICE candidate created for 382789537: media=audio idx=0 name=candidate:1 1 UDP 2122317823 [IP4] [PORT] typ host
ICE candidate created for 382789537: media=audio idx=0 name=candidate:3 1 UDP 1686109695 [IP4] [PORT] typ srflx raddr 0.0.0.0 rport 0
ICE candidate created for 382789537: media=audio idx=0 name=candidate:4 1 UDP 8387839 [IP4] [PORT] typ relay raddr 0.0.0.0 rport 0
```

It feels like I am just missing a small thing here, that would make it work. If you have any more ideas where I could be doing something wrong or missing a step, I would greatly appreciate any pointers and advice.

Thanks in advance for any input you can give on this issue and best regards,
Patrick

**Reply by paullouisageneau** [reply]
If you add the track on the server, `onTrack()` won't be triggered there, it'll be triggered on the client, as it receives the offered track. The server simply uses the object returned by `addTrack()`. It works similarly to `createDataChannel()` and `onDataChannel()`.

**Reply by yrucrem** [reply]
Hello and thanks for your quick answer!

I should have been more explicit in my last post sorry. I see the log messages only for the local events on both ends. But thanks for the hint, I will double check if the client build supresses the output for some reason.

If I understand you correctly, the generated session descriptions should be fine, and the track should actually be created on the other peer connection?

Thanks again for you assistance and best regards,
Patrick

**Reply by paullouisageneau** [reply]
I'm not sure what your setup is, earlier you referred to `onTrack()` on server side. Do you actually mean `onTrack()` on client side?

If both `onTrack()` and `onDataChannel()` are not triggered on answerer side whereas the peer connection reaches connected state, it's probably not a track issue, but possibly user code blocking a callback.

**Reply by yrucrem** [reply]
Yes I mean I don't see the expected log message in `onTrack()` on the client side, sorry for the confusion. The track is added on server side and I understand from your previous answer, that `onTrack()` is not called on the side that adds the track.

**Reply by yrucrem** [reply]
Hello again!

As you suggested I checked again, if user code was blocking the callback and you were right, the callback is called, but on a windows build the console output was suppressed, so I just did not see the messages and I could fix that now.

So I am a step further and get the remote track that was added on the server, thank you again for all your input!

Unfortunatelly I ran into a similar problem, when I try to add a track on both sides of the peer connection (one to send audio data from the server to the client and the other one for the reverse direction): In this case I do not get any calls to the `onTrack()` callbacks on both sides.

The process is as follows:

1. Server creates peer connection.
2. Server adds audio track (ssrc=2, cname="audio-stream")
3. Server creates datachannels.
4. Server calls `setLocalDescription()`
5. Server sends created offer to client via signaling connection.
6. Client creates peer connection.
7. Client adds audio track (ssrc=3, cname="audio-stream")
8. Client sets offer as remote description.
9. Client sets generated answer as local description.
10. Client sends created answer to server.
11. Server sets answer as remote description.

The `disableAutoNeogotiation`  and `forceMediaTransport` settings are true on both sides.

This is the generated offer:

> v=0
o=rtc 2262753625 0 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE audio 0
a=group:LS audio
a=msid-semantic:WMS *
a=setup:actpass
a=ice-ufrag:4HUi
a=ice-pwd:****
a=ice-options:ice2,trickle
a=fingerprint:sha-256 ****
m=audio 9 UDP/TLS/RTP/SAVPF 96
c=IN IP4 0.0.0.0
a=mid:audio
a=sendonly
a=ssrc:2 cname:audio-stream
a=rtcp-mux
a=rtpmap:96 opus/48000/2
a=fmtp:96 minptime=10;maxaveragebitrate=96000;stereo=1;sprop-stereo=1;useinbandfec=1
m=application 9 UDP/DTLS/SCTP webrtc-datachannel
c=IN IP4 0.0.0.0
a=mid:0
a=sendrecv
a=sctp-port:5000
a=max-message-size:262144

and this the generated answer:

> v=0
o=rtc 1109582792 0 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE audio 0
a=group:LS audio
a=msid-semantic:WMS *
a=setup:active
a=ice-ufrag:K7zj
a=ice-pwd:***
a=ice-options:ice2,trickle
a=fingerprint:sha-256 ***
m=audio 9 UDP/TLS/RTP/SAVPF 96
c=IN IP4 0.0.0.0
a=mid:audio
a=sendonly
a=ssrc:3 cname:audio-stream
a=rtcp-mux
a=rtpmap:96 opus/48000/2
a=fmtp:96 minptime=10;maxaveragebitrate=96000;stereo=1;sprop-stereo=1;useinbandfec=1
m=application 9 UDP/DTLS/SCTP webrtc-datachannel
c=IN IP4 0.0.0.0
a=mid:0
a=sendrecv
a=sctp-port:5000
a=max-message-size:262144

I would expect the answer to contain both audio-tracks not just the one created on the client. When only the server adds an audtio-track, I get the log message about the received audio-track directly after setting the offer as remote description.

I suspect that I am doing something I am not supposed to or have the order of the steps mixed up. That's why I added my steps in such detail (please excuse the long post, I did not want to miss any important information).

Do you maybe see any problems with the taken steps and the created session descriptions?

Thanks again for your help and any further assistance, it is greatly appreciated!

Best regards,
Patrick

**Reply by paullouisageneau** [reply]
You can't add a track in an answer by specification, only an offer can contain new tracks. You would need a renegotiation to add a track from the other side.

However, you don't need two tracks in this case, simply create one track with sendrecv direction. The client will need to add the SSRC it is going to send in `onTrack()`:
```cpp
peer_connection->onTrack([](std::shared_ptr<rtc::Track> track) {
    auto description = track->description();
    // [You can check here what direction, payload type, and format the other side offers]
    description.addSSRC(ssrc, "audio-send");
    track->setDescription(description);
});
```
In the general case, a single offer is sufficient provided no track is added after connection, as you can also offer to receive only.

**Reply by yrucrem** [reply]
Ah thank you very much for the clarification, I was suspecting that somehting was not quite right in my usage of the library.

Is there anything special I need to do on the tracks media handlers when I want to send data in both directions? You made me aware, that I need to add the Opus RTP packetizer via `setMediaHandler()`, Is this also necessary when receiving the track in the `onTrack()` callback (with a new rtpConfig containing the other SSRC)?

In the media-receiver example I see that the added media handler is a `RtcpReceivingSession` and I am not sure how to set both handlers.

If I am not mistaken there is no example where a media track is used both for sending and receiving, could you maybe provide a simple example? Or if I just overlooked it point me to an example that covers this topic?

Thanks for your continued support, it really helps a lot to understand this topic better.

Best regards,
Patrick

**Reply by paullouisageneau** [reply, marked as answer]
> Is there anything special I need to do on the tracks media handlers when I want to send data in both directions? You made me aware, that I need to add the Opus RTP packetizer via `setMediaHandler()`, Is this also necessary when receiving the track in the `onTrack()` callback (with a new rtpConfig containing the other SSRC)?

Yes indeed, if you send the same format in both directions, you need to set up the track the same way in `onTrack()`.

> In the media-receiver example I see that the added media handler is a `RtcpReceivingSession` and I am not sure how to set both handlers.
> 
> If I am not mistaken there is no example where a media track is used both for sending and receiving, could you maybe provide a simple example? Or if I just overlooked it point me to an example that covers this topic?

There is no example for bidirectional tracks, but the way you set multiple handlers is the same as for unidirectional tracks, only you need to do it both on `addTrack()` and `onTrack()`. Handlers may be chained together to perform the processing you want. For instance in your case, you should do something like:
```cpp
  auto rtpConfig = make_shared<RtpPacketizationConfig>(ssrc, "audio-send", 96);
  auto handler = std::make_shared<rtc::OpusRtpPacketizer>(rtpConfig); // packetize outgoing
  handler->addToChain(std::make_shared<rtc::OpusRtpDepacketizer>()); // depacketize incoming
  track->setMediaHandler(handler);
```
It will work like this but to make it complete you should also send RTCP reports:
```cpp
  auto rtpConfig = make_shared<RtpPacketizationConfig>(ssrc, "audio-send", 96);
  auto handler = std::make_shared<rtc::OpusRtpPacketizer>(rtpConfig); // packetize outgoing
  handler->addToChain(std::make_shared<rtc::OpusRtpDepacketizer>()); // depacketize incoming
  handler->addToChain(std::make_shared<rtc::RtcpReceivingSession>()); // to send RTCP receiver reports
  handler->addToChain(std::make_shared<rtc::RtcpSrReporter>()); // to send RTCP sender reports
  track->setMediaHandler(handler);
```

To receive, you should use the `onFrame()` callback which gives you metadata like the timestamp contrary to `onMessage()`:
```cpp
  track->onFrame([](const rtc::binary data, rtc::FrameInfo info) {
	// handle the data
  });
```

And thank you so much for the ko-fi!

**Reply by yrucrem** [reply]
Thank you very much, that helped a lot to clear things up! I will try that as soon as I can and give an update if I got it to work.

> And thank you so much for the ko-fi!

You're perfectly welcome, it feels like the least I could do. :-) I appreciate your work on the library and your swift support very much!

**Reply by yrucrem** [reply]
Small update: I did my first tests, and the audio data is correctly transferred between client and server. Thanks again for your support, it helped a lot for understanding libdatachannel und WebRTC intricacies!

**Reply by paullouisageneau** [reply]
Great, you are welcome!

**Reply by ericaaxa** [reply]
why I use OpusRtpDepacketizer   it seems link error
webrtc_client.cpp.obj : error LNK2019: 无法解析的外部符号 "__declspec(dllimport) public: __cdecl rtc::AudioRtpDepacketizer<48000>::AudioRtpDepacketizer<48000>(unsigned int)" (__imp_??0?$AudioRtpDepacketizer@$0LLIA@@rtc@@QEAA@I@Z)，函数 "public: __cdecl std::_Ref_count_obj2<class rtc::AudioRtpDepacketizer<48000> >::_Ref_count_obj2<class rtc::AudioRtpDepacketizer<48000> ><unsigned int const &>(unsigned int const &)" (??$?0AEBI@?$_Ref_count_obj2@V?$AudioRtpDepacketizer@$0LLIA@@rtc@@@std@@QEAA@AEBI@Z) 中引用了该符号
